File: android/src/main/java/me/carda/awesome_notifications/awesome_notifications_core/services/BackgroundService.java
package me.carda.awesome_notifications.awesome_notifications_core.services;

import android.content.Context;
import android.content.Intent;

import me.carda.awesome_notifications.awesome_notifications_core.exceptions.ExceptionCode;
import me.carda.awesome_notifications.awesome_notifications_core.exceptions.ExceptionFactory;
import me.carda.awesome_notifications.awesome_notifications_core.logs.Logger;

import androidx.annotation.NonNull;
import androidx.core.app.JobIntentService;
import me.carda.awesome_notifications.awesome_notifications_core.background.BackgroundExecutor;
import me.carda.awesome_notifications.awesome_notifications_core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.awesome_notifications_core.managers.DefaultsManager;

public class BackgroundService extends JobIntentService {

    private static final String TAG = "BackgroundService";

    @Override
    protected void onHandleWork(@NonNull final Intent intent) {

        Logger.d(TAG, "A new Dart background service has started");
        try {

            Long dartCallbackHandle = getDartCallbackDispatcher(this);
            if (dartCallbackHandle == 0L) {
                ExceptionFactory
                        .getInstance()
                        .registerNewAwesomeException(
                                TAG,
                                ExceptionCode.CODE_BACKGROUND_EXECUTION_EXCEPTION,
                                "A background message could not be handled in Dart" +
                                " because there is no onActionReceivedMethod handler registered.",
                                ExceptionCode.DETAILED_INVALID_ARGUMENTS+".dartCallback");
                return;
            }

            Long silentCallbackHandle = getSilentCallbackDispatcher(this);
            if (silentCallbackHandle == 0L) {
                ExceptionFactory
                        .getInstance()
                        .registerNewAwesomeException(
                                TAG,
                                ExceptionCode.CODE_BACKGROUND_EXECUTION_EXCEPTION,
                                "A background message could not be handled in Dart" +
                                " because there is no dart background handler registered.",
                                ExceptionCode.DETAILED_INVALID_ARGUMENTS+".silentCallback");
                return;
            }

            BackgroundExecutor.runBackgroundExecutor(
                    this,
                    intent,
                    dartCallbackHandle,
                    silentCallbackHandle);

        } catch (AwesomeNotificationsException ignored) {
        } catch (Exception e) {
            ExceptionFactory
                    .getInstance()
                    .registerNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_BACKGROUND_EXECUTION_EXCEPTION,
                            "A new Dart background service could not be executed",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS,
                            e);
        }
    }

    public static Long getDartCallbackDispatcher(Context context) throws AwesomeNotificationsException {
        return DefaultsManager.getDartCallbackDispatcher(context);
    }

    public static Long getSilentCallbackDispatcher(Context context) throws AwesomeNotificationsException {
        return DefaultsManager.getSilentCallbackDispatcher(context);
    }
}

File: ios/AwnCore/Classes/services/BackgroundService.swift
//
//  BackgroundService.swift
//  awesome_notifications
//
//  Created by CardaDev on 02/02/22.
//

import Foundation

class BackgroundService {
    
    static let TAG = "BackgroundService"
    
    // ************** SINGLETON PATTERN ***********************
    
    static var instance:BackgroundService?
    public static var shared:BackgroundService {
        get {
            BackgroundService.instance =
                BackgroundService.instance ?? BackgroundService()
            return BackgroundService.instance!
        }
    }
    private init(){}
    
    // ********************************************************
    
    public func enqueue(
        SilentBackgroundAction silentAction: ActionReceived,
        withCompletionHandler completionHandler: @escaping (Bool, Error?) -> ()
    ){
        let start = DispatchTime.now()
        Logger.d(BackgroundService.TAG, "A new Dart background service has started")
        
        let completionWithTimer:(Bool, Error?) -> () = { (success, error) in
            let end = DispatchTime.now()
            let nanoTime = end.uptimeNanoseconds - start.uptimeNanoseconds
            let timeInterval:Double = Double(nanoTime) / 1_000_000
            Logger.d(BackgroundService.TAG, "Background action finished in \(timeInterval.rounded())ms")
            
            completionHandler(success, error)
        }
        
        do {
            let backgroundCallback:Int64 = DefaultsManager.shared.backgroundCallback
            let actionCallback:Int64 = DefaultsManager.shared.actionCallback
            
            if backgroundCallback == 0 {
                throw ExceptionFactory
                        .shared
                        .createNewAwesomeException(
                            className: BackgroundService.TAG,
                            code: ExceptionCode.CODE_INVALID_ARGUMENTS,
                            message: "A background message could not be handled in Dart because there is no valid background handler register",
                            detailedCode: ExceptionCode.DETAILED_INVALID_ARGUMENTS + ".enqueue.backgroundCallback")
            }
            
            if actionCallback == 0 {
                throw ExceptionFactory
                        .shared
                        .createNewAwesomeException(
                            className: BackgroundService.TAG,
                            code: ExceptionCode.CODE_INVALID_ARGUMENTS,
                            message: "A background message could not be handled in Dart because there is no valid action callback handler register",
                            detailedCode: ExceptionCode.DETAILED_INVALID_ARGUMENTS + ".enqueue.backgroundCallback")
            }
            
            if Thread.isMainThread {
                mainThreadServiceExecution(
                    SilentBackgroundAction: silentAction,
                    backgroundCallback: backgroundCallback,
                    actionCallback: actionCallback,
                    withCompletionHandler: completionWithTimer)
            }
            else {
                try backgroundThreadServiceExecution(
                    SilentBackgroundAction: silentAction,
                    backgroundCallback: backgroundCallback,
                    actionCallback: actionCallback,
                    withCompletionHandler: completionWithTimer)
            }
            
        } catch {
            if error is AwesomeNotificationsException {
                completionWithTimer(false, error)
            } else {
                completionWithTimer(
                    false,
                    ExceptionFactory
                        .shared
                        .createNewAwesomeException(
                            className: BackgroundService.TAG,
                            code: ExceptionCode.CODE_INVALID_ARGUMENTS,
                            message: "A background message could not be handled in Dart because there is no valid background handler register",
                            detailedCode: ExceptionCode.DETAILED_INVALID_ARGUMENTS + ".enqueue.backgroundCallback"))
            }
        }
    }
    
    func mainThreadServiceExecution(
        SilentBackgroundAction silentAction: ActionReceived,
        backgroundCallback:Int64,
        actionCallback:Int64,
        withCompletionHandler completionHandler: @escaping (Bool, Error?) -> ()
    ){
        let silentActionRequest:SilentActionRequest =
                SilentActionRequest(
                    actionReceived: silentAction,
                    handler: { success in
                        completionHandler(success, nil)
                    })
        
        let backgroundExecutor:BackgroundExecutor =
                AwesomeNotifications
                    .backgroundClassType!
                    .init()
        
        backgroundExecutor
            .runBackgroundProcess(
                silentActionRequest: silentActionRequest,
                dartCallbackHandle: backgroundCallback,
                silentCallbackHandle: actionCallback)
    }
    
    func backgroundThreadServiceExecution(
        SilentBackgroundAction silentAction: ActionReceived,
        backgroundCallback:Int64,
        actionCallback:Int64,
        withCompletionHandler completionHandler: @escaping (Bool, Error?) -> ()
    ) throws {
        let group = DispatchGroup()
        group.enter()
        
        let silentActionRequest:SilentActionRequest =
                SilentActionRequest(
                    actionReceived: silentAction,
                    handler: { success in
                        group.leave()
                    })
        
        let workItem:DispatchWorkItem = DispatchWorkItem {
            DispatchQueue.global(qos: .background).async {
                
                let backgroundExecutor:BackgroundExecutor =
                        AwesomeNotifications
                            .backgroundClassType!
                            .init()
                
                backgroundExecutor
                    .runBackgroundProcess(
                        silentActionRequest: silentActionRequest,
                        dartCallbackHandle: backgroundCallback,
                        silentCallbackHandle: actionCallback)
            }
            
        }
        
        workItem.perform()
        if group.wait(timeout: DispatchTime.now() + .seconds(10)) == .timedOut {
            workItem.cancel()
            throw ExceptionFactory
                    .shared
                    .createNewAwesomeException(
                        className: BackgroundService.TAG,
                        code: ExceptionCode.CODE_INVALID_ARGUMENTS,
                        message: "Background silent push service reached timeout limit",
                        detailedCode: ExceptionCode.DETAILED_INVALID_ARGUMENTS + ".mainThreadServiceExecution.timeout")
        }
    }
    
}
